#### Pandas

In this notebook we are going to learn `pandas` basics when working with data. We are going to look at the following:

1. Input Output (I/O)🆗
2. Pandas series
3. DataFrames
4. Pandas Arrays
5. Index Objects


If you do not have `pandas` installed you can run the command:

> !pip install pandas

Note that you are using `anaconda ` you don't need to install pandas as it comes installed. So first thing we are going to install `pandas` and check it version thereafter.


In [5]:
import numpy as np
import pandas as pd

#### Input/Output

In this section we are going to look at `io` in pandas

- `CSV` files - these are files columns of data are seperated using a comma, that why it's called `Comma Seperated Values`. We are going to read and write to csv files.

> To read a `csv` file in pandas we use the `.read_csv` function. This function returns us a `pandas` dataframe.

In [8]:
dataframe = pd.read_csv('files/airquality.csv')

In the `dataframe` object we can basically check the first `5` rows of data using the `head()` function.

In [10]:
dataframe.head()

Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
0           1   41.0    190.0   7.4    67      5    1
1           2   36.0    118.0   8.0    72      5    2
2           3   12.0    149.0  12.6    74      5    3
3           4   18.0    313.0  11.5    62      5    4
4           5    NaN      NaN  14.3    56      5    5

In the `dataframe` object we can basically check the last `5` rows of data using the `tail()` function.

In [12]:
dataframe.tail()

Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
148         149   30.0    193.0   6.9    70      9   26
149         150    NaN    145.0  13.2    77      9   27
150         151   14.0    191.0  14.3    75      9   28
151         152   18.0    131.0   8.0    76      9   29
152         153   20.0    223.0  11.5    68      9   30

In the `head()` or `tail()` function you can specify the number of rows you want.

In [14]:
dataframe.head(10)

Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
0           1   41.0    190.0   7.4    67      5    1
1           2   36.0    118.0   8.0    72      5    2
2           3   12.0    149.0  12.6    74      5    3
3           4   18.0    313.0  11.5    62      5    4
4           5    NaN      NaN  14.3    56      5    5
5           6   28.0      NaN  14.9    66      5    6
6           7   23.0    299.0   8.6    65      5    7
7           8   19.0     99.0  13.8    59      5    8
8           9    8.0     19.0  20.1    61      5    9
9          10    NaN    194.0   8.6    69      5   10

In [15]:
dataframe.tail(7)

Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
146         147    7.0     49.0  10.3    69      9   24
147         148   14.0     20.0  16.6    63      9   25
148         149   30.0    193.0   6.9    70      9   26
149         150    NaN    145.0  13.2    77      9   27
150         151   14.0    191.0  14.3    75      9   28
151         152   18.0    131.0   8.0    76      9   29
152         153   20.0    223.0  11.5    68      9   30

Note that all the data that we are working with comes inform of `csv` files. We might also have files like:
    
    1. excel spreadsheet
    2. text file (.txt)
    3. json files
    4. sql files
    5. etc.
    

The `read_excel` method is used to read an excel spreadsheet into a `pandas` dataframe. In our files directory we have an excel file that is called `cars.xlsx` and we want to read this file in a pandas dataframe. We can do it as follows.

In [18]:
df  =  pd.read_excel('files/cars.xlsx')
df.head()

mpg  engine  horse  weight  accel  year  origin  cylinder  filter_.   mpg1
0   9.0     4.0   93.0     732    8.5     0     NaN       NaN       NaN   9.98
1  10.0   360.0  215.0    4615   14.0    70     1.0       8.0       0.0  10.87
2  10.0   307.0  200.0    4376   15.0    70     1.0       8.0       0.0   9.63
3  11.0   318.0  210.0    4382   13.5    70     1.0       8.0       0.0  12.12
4  11.0   429.0  208.0    4633   11.0    72     1.0       8.0       0.0  10.63

> You can use some dataframe methonds like `head()`, `tail()` etc on the `df` object because we know that `read_excel` returns us a pandas dataframe. 

Pandas exposed a method called `read_json()` which allows us to read `json` data `json` file into a pandas dataframe. We have a `json` file that is located in the `files` directory and it's name is `todos.json`. Let's read this file to a pandas dataframe.

In [20]:
df =  pd.read_json('files/todos.json')
df.head(6)

userId  id                                              title  completed
0       1   1                                 delectus aut autem      False
1       1   2                 quis ut nam facilis et officia qui      False
2       1   3                                fugiat veniam minus      False
3       1   4                                   et porro tempora       True
4       1   5  laboriosam mollitia et enim quasi adipisci qui...      False
5       1   6  qui ullam ratione quibusdam voluptatem quia omnis      False

In `pandas` we can read 'txt' files using the `read_csv`. When reading text files we need to specify the delimeter for our columns in the `read_csv` method. We have a file called 'files/airqualify.txt' that contains the data that looks as follows:

txt

Ozone   Solar.R Wind    Temp    Month   Day
41  190 7.4 67  5   1
36  118 8   72  5   2
12  149 12.6    74  5    3
18  313 11.5    62  5    4


> We will read this files into a pandas dataframe using the 'read_csv' and specify the delimeter to be a `\t` because the columns are separated by tabs.

In [25]:
# df = pd.read_csv('files/airquality.txt', delimeter = '\t')

# df.head()

### DataFrame

In this section we are going to have a look at how we can create our own dataframe from iterable using the `Dataframe` class. Let's start by creating a dataframe from `list` of `dictionaries` as follows:


In [31]:
users = [    
    {'gender': 'male', 'name': 'Jonh', 'age': 56, 'surname': 'Doe'},
    {'gender': 'female', 'name': 'Mary', 'age': 26, 'surname': 'Jack'},
    {'gender': 'male', 'name': 'Peter', 'age': 34, 'surname': 'Gross'}
     ]
   
df = pd.DataFrame(users)
df

gender   name  age surname
0    male   Jonh   56     Doe
1  female   Mary   26    Jack
2    male  Peter   34   Gross

We can also use `list` of `lists` to create a pandas dataframe. Let's have a look at the following example.
> When using `list` of `lists` we need to specify the columns which are the column names.


In [34]:
rows = [user.values() for user in users]
headers = users[0].keys()
headers, rows

(dict_keys(['gender', 'name', 'age', 'surname']),
 [dict_values(['male', 'Jonh', 56, 'Doe']),
  dict_values(['female', 'Mary', 26, 'Jack']),
  dict_values(['male', 'Peter', 34, 'Gross'])])

In [35]:
df = pd.DataFrame(rows, columns = headers)
df

gender   name  age surname
0    male   Jonh   56     Doe
1  female   Mary   26    Jack
2    male  Peter   34   Gross

The dataframe object has various a methods of saving data to files. Let's save our dataframes to files in a folder called `saved`.

In [37]:
import os 

if not os.path.exists('saved'):
    os.mkdir('saved')

Let's save our dataframe to a `.csv` file in the folder `saved`.

- Specifying that `index=False` we are telling pandas that we don't want a column of index in our `csv` file.

In [38]:
df.to_csv('saved/people.csv', index=False)
print('Saved')

Saved


We can use the `to_excel` to convert a dataframe to an excel spreadsheet for example

In [39]:
df.to_excel('saved/people.xlsx', index=False)
print('Saved')

Saved


We can use the `to_json` to convert a dataframe to a `json` file for exmaple. We need to specify the `orient`.

If we do not want our `json` file to have a list of dictionaries we need to set the lines to be `False` else the output in the file will be `lines` of dictionaries.

In [40]:
df.to_json('saved/people.json', orient='records', lines=False, indent=2)
print('Saved')

Saved


### DataFrame object 

Here we are going to look at some methonds we can use on the `DataFrame` class. 

* We have a method called `reset_index()`. This method is responsible for resteing the index in a dataframe object.

In [41]:
dataframe.tail(3).reset_index(drop = True)

Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
0         151   14.0    191.0  14.3    75      9   28
1         152   18.0    131.0   8.0    76      9   29
2         153   20.0    223.0  11.5    68      9   30

In [42]:
dataframe.tail(3)


Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
150         151   14.0    191.0  14.3    75      9   28
151         152   18.0    131.0   8.0    76      9   29
152         153   20.0    223.0  11.5    68      9   30

###### NB 

We can access the `columns` of a dataframe by calling the `columns property`

In [44]:
columns = dataframe.columns
columns

Index(['Unnamed: 0', 'Ozone', 'Solar.R', 'Wind', 'Temp', 'Month', 'Day'], dtype='object')

We can get the values using the `values` property.

In [45]:
dataframe.values

array([[  1.,  41., 190., ...,  67.,   5.,   1.],
       [  2.,  36., 118., ...,  72.,   5.,   2.],
       [  3.,  12., 149., ...,  74.,   5.,   3.],
       ...,
       [151.,  14., 191., ...,  75.,   9.,  28.],
       [152.,  18., 131., ...,  76.,   9.,  29.],
       [153.,  20., 223., ...,  68.,   9.,  30.]])

In pandas dataframe we can access the values in the `column` that we are intrested in using the `values` propety. But first we need to index the column name that we want data from first.

* Calling values will return a numpy version of values. Calling `dataframe['Month']` or ` dataframe.Month` returns us a pandas [Series](https://pandas.pydata.org/docs/reference/api/pandas.Series.html#pandas.Series), inorder for us to get values in form of numpy we need to call the .values for example:



In [46]:
months = dataframe['Month'].values
months

array([5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
       5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
       6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 7, 7, 7, 7,
       7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
       7, 7, 7, 7, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8,
       8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 9, 9, 9, 9, 9, 9, 9, 9, 9,
       9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9],
      dtype=int64)

In [47]:
# Or
months = dataframe.Month.values
months

array([5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
       5, 5, 5, 5, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,
       6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 7, 7, 7, 7, 7,
       7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7,
       7, 7, 7, 7, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8,
       8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 9, 9, 9, 9, 9, 9, 9, 9, 9,
       9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9],
      dtype=int64)

We can call the `unique()` method on a pandas series to get the unique values of that column

In [48]:
months = dataframe.Month.unique()
months

array([5, 6, 7, 8, 9], dtype=int64)

In [49]:
dataframe.head()

Unnamed: 0  Ozone  Solar.R  Wind  Temp  Month  Day
0           1   41.0    190.0   7.4    67      5    1
1           2   36.0    118.0   8.0    72      5    2
2           3   12.0    149.0  12.6    74      5    3
3           4   18.0    313.0  11.5    62      5    4
4           5    NaN      NaN  14.3    56      5    5

Let's say we want to get the first `5` temperatures or last `5` temperatures in our dataframe. We can do it as follows:

In [52]:
dataframe.Temp.head()
# dataframe['Temp'].head() one and the same thing

0    67
1    72
2    74
3    62
4    56
Name: Temp, dtype: int64

In [55]:
dataframe['Temp'].tail(2).reset_index(drop=True)

0    76
1    68
Name: Temp, dtype: int64

In our dataframe we have a column called `Unnamed: 0` Let's go ahead and drop or remove that column from our dataframe. We can use the `drop` method. In the drop method we can specify a kwarg called `inplace` by default it is set to `False`. Setting it to `True` will modify the dataframe inplace.

In [56]:
dataframe.drop(columns=['Unnamed: 0'], inplace=True)

In [57]:
dataframe.head()

Ozone  Solar.R  Wind  Temp  Month  Day
0   41.0    190.0   7.4    67      5    1
1   36.0    118.0   8.0    72      5    2
2   12.0    149.0  12.6    74      5    3
3   18.0    313.0  11.5    62      5    4
4    NaN      NaN  14.3    56      5    5

~~In our dataframe we have a column called Unnamed: 0 Let's go ahead and drop or remove that column from our dataframe. We can use the drop method. In the drop method we can specify a kwarg called inplace by default it i~~s set to ~~~~~~False. ~~~~ ~~Setting ~~it to True will modify the dataframe inplace~~~~.

We can select columns that we want in a dataframe using `indexing`. Let's demostrate by sellecting ` Month` and `Temp` columns from our dataframe.

In [61]:
dataframe[['Month','Temp']].head()

Month  Temp
0      5    67
1      5    72
2      5    74
3      5    62
4      5    56

If we want to get the rows we can use `loc` or `iloc` methods to index rows.

In [62]:
dataframe.iloc[3]

Ozone       18.0
Solar.R    313.0
Wind        11.5
Temp        62.0
Month        5.0
Day          4.0
Name: 3, dtype: float64

In [63]:
#we can also use slicing to get the 3 rows as follows:
dataframe.iloc[:3]

Ozone  Solar.R  Wind  Temp  Month  Day
0   41.0    190.0   7.4    67      5    1
1   36.0    118.0   8.0    72      5    2
2   12.0    149.0  12.6    74      5    3

We can create a column based on another column in a dataframe. We want to create a column called `Degrees`. In this column we want to take the `Temp` column and convert it from `F ` to `C`.

In [64]:
dataframe['Degrees'] = (dataframe['Temp']-32) * 5/9
dataframe.head()

Ozone  Solar.R  Wind  Temp  Month  Day    Degrees
0   41.0    190.0   7.4    67      5    1  19.444444
1   36.0    118.0   8.0    72      5    2  22.222222
2   12.0    149.0  12.6    74      5    3  23.333333
3   18.0    313.0  11.5    62      5    4  16.666667
4    NaN      NaN  14.3    56      5    5  13.333333

We have a attribute called `at` which allows us to index the value at a specified `row` and at a specified `column` name. Let's access the value `23.333333` from our dataframe. We can do it as follows:

In [65]:
dataframe.at[2,'Degrees']

23.333333333333332

We can use the `iat` attribute to acess the value using the `row` index and `column` index.

In [67]:
dataframe.iat[2,6]

23.333333333333332

We can check the datatypes for each column in a dataframe using the `dtypes` as follows:

In [68]:
dataframe.dtypes

Ozone      float64
Solar.R    float64
Wind       float64
Temp         int64
Month        int64
Day          int64
Degrees    float64
dtype: object

We can also check the shape of the dataframe using the `shape` attribute as follows:

In [70]:
dataframe.shape # 153 rows and 7 columns

(153, 7)

We can check the size of the dataframe using the `size` property as follows.

In [71]:
dataframe.size

1071

We can check for `null` values in colums using the `isnull()` function as follows.

In [72]:
dataframe.isnull().any()

Ozone       True
Solar.R     True
Wind       False
Temp       False
Month      False
Day        False
Degrees    False
dtype: bool

We can check if we have `na` values using `isna()` function as follows:

In [73]:
dataframe.isna().any()

Ozone       True
Solar.R     True
Wind       False
Temp       False
Month      False
Day        False
Degrees    False
dtype: bool

We can clean the dataframe by droping all the columns that have `null values/ na values`  using the `dropna` method. We can do it in two ways:

    1. `dataframe = dataframe.dropna()
    2. dataframe.dropna(inplace = True)`

In [76]:
dataframe.dropna(inplace=True)
dataframe.head()

Ozone  Solar.R  Wind  Temp  Month  Day    Degrees
0   41.0    190.0   7.4    67      5    1  19.444444
1   36.0    118.0   8.0    72      5    2  22.222222
2   12.0    149.0  12.6    74      5    3  23.333333
3   18.0    313.0  11.5    62      5    4  16.666667
6   23.0    299.0   8.6    65      5    7  18.333333

The are core methods that we can work with in pandas dataframes. Let's have a look at the `abs()` method on a dataframe.

In [78]:
dataframe.abs().head(2) #this one work if you want to convert negative values to +


Ozone  Solar.R  Wind  Temp  Month  Day    Degrees
0   41.0    190.0   7.4    67      5    1  19.444444
1   36.0    118.0   8.0    72      5    2  22.222222

We can also use methods like:
1. add()
2. div()
3. mod()
4. etc.

[on the dataframe columns](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.add.html)

In [79]:
dataframe.head().add(2) #add 2 to each entry 

Ozone  Solar.R  Wind  Temp  Month  Day    Degrees
0   43.0    192.0   9.4    69      7    3  21.444444
1   38.0    120.0  10.0    74      7    4  24.222222
2   14.0    151.0  14.6    76      7    5  25.333333
3   20.0    315.0  13.5    64      7    6  18.666667
6   25.0    301.0  10.6    67      7    9  20.333333

We can create a copy of a dataframe using the `copy()` method as follows:

In [80]:

df3 = dataframe.copy()
df3.head()

Ozone  Solar.R  Wind  Temp  Month  Day    Degrees
0   41.0    190.0   7.4    67      5    1  19.444444
1   36.0    118.0   8.0    72      5    2  22.222222
2   12.0    149.0  12.6    74      5    3  23.333333
3   18.0    313.0  11.5    62      5    4  16.666667
6   23.0    299.0   8.6    65      5    7  18.333333

Let's say we are given a task that says we should create a column `isHot` which checks the temperature in `Degrees` if it is greater that `20 or not`.

In [81]:
dataframe['isHot'] = np.where(dataframe.Degrees > 20, 'Hot', 'Cold')
dataframe.head()


Ozone  Solar.R  Wind  Temp  Month  Day    Degrees isHot
0   41.0    190.0   7.4    67      5    1  19.444444  Cold
1   36.0    118.0   8.0    72      5    2  22.222222   Hot
2   12.0    149.0  12.6    74      5    3  23.333333   Hot
3   18.0    313.0  11.5    62      5    4  16.666667  Cold
6   23.0    299.0   8.6    65      5    7  18.333333  Cold

We can get the values rows of our dataframe based on a condition. Let's get all the rows of data where the isHot value is Hot

In [82]:
hot_df = dataframe[dataframe.isHot == 'Hot']
hot_df.head()

Ozone  Solar.R  Wind  Temp  Month  Day    Degrees isHot
1    36.0    118.0   8.0    72      5    2  22.222222   Hot
2    12.0    149.0  12.6    74      5    3  23.333333   Hot
11   16.0    256.0   9.7    69      5   12  20.555556   Hot
21   11.0    320.0  16.6    73      5   22  22.777778   Hot
28   45.0    252.0  14.9    81      5   29  27.222222   Hot

In the file called `HairEyeColor.csv` we want to read the data that is in that file and ge all the people that has black hear and whose gender is Male

In [83]:

df = pd.read_csv("files/HairEyeColor.csv")
df.drop(columns=['Unnamed: 0'], inplace=True)
df.head()

Hair    Eye   Sex  Freq
0  Black  Brown  Male    32
1  Brown  Brown  Male    53
2    Red  Brown  Male    10
3  Blond  Brown  Male     3
4  Black   Blue  Male    11

We can use the `loc` method as follows:

In [84]:
df.loc[(df.Sex == 'Male') & (df.Hair == 'Black')]

Hair    Eye   Sex  Freq
0   Black  Brown  Male    32
4   Black   Blue  Male    11
8   Black  Hazel  Male    10
12  Black  Green  Male     3

Or you can do it as follows:

In [85]:
df[(df.Sex == 'Male') & (df.Hair == 'Black')]

Hair    Eye   Sex  Freq
0   Black  Brown  Male    32
4   Black   Blue  Male    11
8   Black  Hazel  Male    10
12  Black  Green  Male     3


Next we are going to concatenate the dataframe based on a condition. What you should note is that:

1. Rows must be equal if we are adding new colums
2. Column values must be equal to the one that exists if we are adding new rows.

In [86]:
A = dataframe.head(10)
B = df.head(10)

Let's concatenate `A` and `B` based on column. For that we should specify the `axis=1`

In [87]:
pd.concat([A, B], axis=1)

Ozone  Solar.R  Wind  Temp  Month   Day    Degrees isHot   Hair    Eye  \
0    41.0    190.0   7.4  67.0    5.0   1.0  19.444444  Cold  Black  Brown   
1    36.0    118.0   8.0  72.0    5.0   2.0  22.222222   Hot  Brown  Brown   
2    12.0    149.0  12.6  74.0    5.0   3.0  23.333333   Hot    Red  Brown   
3    18.0    313.0  11.5  62.0    5.0   4.0  16.666667  Cold  Blond  Brown   
6    23.0    299.0   8.6  65.0    5.0   7.0  18.333333  Cold    Red   Blue   
7    19.0     99.0  13.8  59.0    5.0   8.0  15.000000  Cold  Blond   Blue   
8     8.0     19.0  20.1  61.0    5.0   9.0  16.111111  Cold  Black  Hazel   
11   16.0    256.0   9.7  69.0    5.0  12.0  20.555556   Hot    NaN    NaN   
12   11.0    290.0   9.2  66.0    5.0  13.0  18.888889  Cold    NaN    NaN   
13   14.0    274.0  10.9  68.0    5.0  14.0  20.000000  Cold    NaN    NaN   
4     NaN      NaN   NaN   NaN    NaN   NaN        NaN   NaN  Black   Blue   
5     NaN      NaN   NaN   NaN    NaN   NaN        NaN   NaN  Brown   Blue   
9     NaN      NaN   NaN   NaN    NaN   NaN        NaN   NaN  Brown  Hazel   

     Sex  Freq  
0   Male  32.0  
1   Male  53.0  
2   Male  10.0  
3   Male   3.0  
6   Male  10.0  
7   Male  30.0  
8   Male  10.0  
11   NaN   NaN  
12   NaN   NaN  
13   NaN   NaN  
4   Male  11.0  
5   Male  50.0  
9   Male  25.0

Let's say we want to concatenate A and B based on a condition that the `isHot` column value in A is Cold we can do it in numerous steps for simplicity as follows:

1. concatenate A and B
2. create a dumy colum condition that set the value of the column to either True or False
3. filter the rows that have column value False
4. Drop the dummy column condition.

In [88]:
A_B = pd.concat([A, B], axis=1)
A_B['condition'] = np.where(A_B.isHot == 'Cold', True, False)
A_B = A_B.loc[A_B.condition == True]
A_B.drop(columns=['condition'], inplace=True)
A_B.head()

Ozone  Solar.R  Wind  Temp  Month  Day    Degrees isHot   Hair    Eye  \
0   41.0    190.0   7.4  67.0    5.0  1.0  19.444444  Cold  Black  Brown   
3   18.0    313.0  11.5  62.0    5.0  4.0  16.666667  Cold  Blond  Brown   
6   23.0    299.0   8.6  65.0    5.0  7.0  18.333333  Cold    Red   Blue   
7   19.0     99.0  13.8  59.0    5.0  8.0  15.000000  Cold  Blond   Blue   
8    8.0     19.0  20.1  61.0    5.0  9.0  16.111111  Cold  Black  Hazel   

    Sex  Freq  
0  Male  32.0  
3  Male   3.0  
6  Male  10.0  
7  Male  30.0  
8  Male  10.0

# References 

[pandas.pydata.org](https://pandas.pydata.org/docs/user_guide/index.html)

******************Banele_of_@astro**********************